File: go-sdk-advanced/codec/bool_codec.go
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package codec

import "fmt"

type BoolCodec struct{}

func (c BoolCodec) Encode(value bool) ([]byte, error) {
	if value {
		return []byte{1}, nil
	}
	return []byte{0}, nil
}

func (c BoolCodec) Decode(data []byte) (bool, error) {
	if len(data) != 1 {
		return false, fmt.Errorf("invalid bool payload length: %d", len(data))
	}
	switch data[0] {
	case 0:
		return false, nil
	case 1:
		return true, nil
	default:
		return false, fmt.Errorf("invalid bool payload byte: %d", data[0])
	}
}

func (c BoolCodec) EncodedSize() int { return 1 }

func (c BoolCodec) IsOrderedKeyCodec() bool { return true }

File: python/functionstream-api-advanced/src/fs_api_advanced/codec/bool_codec.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .base import Codec


class BoolCodec(Codec[bool]):
    supports_ordered_keys = True
    _SIZE = 1

    def encoded_size(self) -> int:
        return self._SIZE

    def encode(self, value: bool) -> bytes:
        return b"\x01" if value else b"\x00"

    def decode(self, data: bytes) -> bool:
        if len(data) != 1:
            raise ValueError(f"invalid bool payload length: {len(data)}")
        if data == b"\x00":
            return False
        if data == b"\x01":
            return True
        raise ValueError(f"invalid bool payload byte: {data[0]}")
